pylint/checkers/unicode.py

```diff
@@ -0,0 +1,701 @@
+                "of this issue. And "
+                "https://softwareengineering.stackexchange.com/questions/102205/should-utf-16-be-considered-harmful"
+                "for some possible problems when using UTF-16 for instance."
+            ),
+        ),
+        "E2402": (
+            "bidirectional unicode can be used to hide real code execution",
+            "bidirectional-unicode",
+            "bidirectional unicode can be used to hide real code execution",
+        ),
+        **{
+            bad_char.code: (
+                bad_char.description(),
+                bad_char.human_code(),
+                bad_char.help_text,
+            )
+            for bad_char in _BAD_CHARS
+        },
+    }
+
+    @staticmethod
+    def _is_invalid_codec(codec: Optional[str]) -> bool:
+        if codec is None:
+            return False
+        codec = codec.lower().replace("-", "").replace("_", "")
+        return codec.startswith("utf16") or codec.startswith("utf32")
+
+    @classmethod
+    def _find_line_matches(cls, line: bytes, codec: str) -> Dict[int, _BadChar]:
+        """Find all matches of BAD_CHARS within line
+
+        Args:
+            line: the input
+            codec: that will be used to convert line/or search string into
+
+        Return:
+            An dictionary with the column offset and the BadASCIIChar
+        """
+        # We try to decode in Unicode to get the correct column offset
+        # if we would use bytes, it could be off because UTF-8 has no fixed length
+        try:
+            line_search = line.decode(codec, errors="strict")
+            search_dict = cls._BAD_ASCII_SEARCH_DICT
+            # If everything is correctly decoded, the BOM will be removed automatically
+            return _map_positions_to_result(line_search, search_dict, "\n")
+        except UnicodeDecodeError:
+            # If we can't decode properly, we simply use bytes, even so the column offsets
+            # might be wrong a bit, but it is still better then nothing
+            line_search_byte = line
+            search_dict_byte = {
+                _cached_encode_search(char.unescaped, codec): char
+                for char in cls._BAD_CHARS
+            }
+            # Using bytes we have to remove the BOM manually
+            return _map_positions_to_result(
+                line_search_byte,
+                search_dict_byte,
+                _cached_encode_search("\n", codec),
+                byte_str_length=_byte_to_str_length(codec),
+            )
+
+    def process_module(self, node: nodes.Module) -> None:
+        """Perform the actual check by checking module stream
+        the module's content is accessible via node.stream() function
+        """
+
+        # Check for UTF-16/UTF-32, to warn for it
+        lineno: int
+        line: bytes
+        codec: Optional[str]
+        lines: Sequence[bytes]
+        invalid_codec_line: Optional[int] = None
```

Comment:
```suggestion
        invalid_codec_line: Optional[int] = None
```

Reply:
The `Optional[int]` might even be unnecessary, see previous example. Please check this yourself.

Reply:
Again `Explicit is better than implicit.`
For me all this lines help a big deal understanding whats comming next.
Removing codec with optional for sure makes problems, because the first assigment to codec is only `str`.
So I would like to keep it if thats okay.

Reply:
I think the `codec` is a similar case as described above where `mypy` infers that after a potential `None` assignment it changes the value of the variable to `Optional[str]`.

Reply:
These lines do not exist anymore in the current implementation.